File: apps/mobile-ios/SocialImperialism/Core/APIClient.swift
import Foundation

enum APIError: LocalizedError {
    case unauthorized
    case server(String)
    case network(Error)

    var errorDescription: String? {
        switch self {
        case .unauthorized: return "Session expired. Please sign in again."
        case .server(let msg): return msg
        case .network(let err): return err.localizedDescription
        }
    }
}

final class APIClient {
    static let shared = APIClient()
    static let productionBase = "https://api.socialimperialism.com"

    var baseURL: String {
        ProcessInfo.processInfo.environment["SI_API_URL"] ?? Self.productionBase
    }

    private let session: URLSession
    private let decoder = JSONDecoder()

    init(session: URLSession = .shared) {
        self.session = session
    }

    // MARK: - Auth

    func login(email: String, password: String) async throws -> SessionResponse {
        try await postJSON("/api/auth/login", body: ["email": email, "password": password], auth: false)
    }

    func register(email: String, password: String, name: String?) async throws -> SessionResponse {
        var body: [String: String] = ["email": email, "password": password]
        if let name, !name.isEmpty { body["name"] = name }
        return try await postJSON("/api/auth/register", body: body, auth: false)
    }

    func me() async throws -> MeResponse {
        try await getJSON("/api/auth/me")
    }

    // MARK: - Invoke

    func invoke<T: Decodable>(_ channel: String, args: [Any] = []) async throws -> T {
        let envelope: InvokeEnvelope<T> = try await postJSON(
            "/api/invoke/\(channel)",
            body: ["args": args],
            auth: true
        )
        guard let data = envelope.data else {
            throw APIError.server("Empty response from \(channel)")
        }
        return data
    }

    func invokeRaw(_ channel: String, args: [Any] = []) async throws -> [String: Any] {
        let data = try await request(
            path: "/api/invoke/\(channel)",
            method: "POST",
            body: ["args": args],
            auth: true
        )
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw APIError.server("Invalid JSON")
        }
        if let checkout = json["checkoutUrl"] as? String {
            await MainActor.run {
                NotificationCenter.default.post(name: .siOpenCheckout, object: checkout)
            }
        }
        return json
    }

    // MARK: - HTTP

    private func getJSON<T: Decodable>(_ path: String) async throws -> T {
        let data = try await request(path: path, method: "GET", body: nil, auth: true)
        return try decode(T.self, from: data)
    }

    private func postJSON<T: Decodable>(_ path: String, body: [String: Any], auth: Bool) async throws -> T {
        let data = try await request(path: path, method: "POST", body: body, auth: auth)
        return try decode(T.self, from: data)
    }

    private func request(path: String, method: String, body: [String: Any]?, auth: Bool) async throws -> Data {
        guard let url = URL(string: baseURL + path) else {
            throw APIError.server("Invalid URL")
        }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if auth {
            guard let token = AuthStore.shared.token else { throw APIError.unauthorized }
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            if let projectId = AuthStore.shared.projectId {
                req.setValue(projectId, forHTTPHeaderField: "x-project-id")
            }
        }

        if let body {
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }

        do {
            let (data, response) = try await session.data(for: req)
            guard let http = response as? HTTPURLResponse else {
                throw APIError.server("No response")
            }
            if http.statusCode == 401 {
                await MainActor.run { AuthStore.shared.clearSession() }
                throw APIError.unauthorized
            }
            if http.statusCode >= 400 {
                let json = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
                let msg = (json?["error"] as? String) ?? HTTPURLResponse.localizedString(forStatusCode: http.statusCode)
                if auth, msg.lowercased().contains("project not found") {
                    AuthStore.shared.projectId = nil
                    _ = try? await AuthStore.shared.repairSession()
                    return try await request(path: path, method: method, body: body, auth: auth)
                }
                throw APIError.server(msg)
            }
            return data
        } catch let err as APIError {
            throw err
        } catch {
            throw APIError.network(error)
        }
    }

    private func decode<T: Decodable>(_ type: T.Type, from data: Data) throws -> T {
        if let wrapper = try? decoder.decode(APIWrapper<T>.self, from: data), let inner = wrapper.data {
            return inner
        }
        return try decoder.decode(T.self, from: data)
    }
}

extension Notification.Name {
    static let siOpenCheckout = Notification.Name("siOpenCheckout")
}

struct APIWrapper<T: Decodable>: Decodable {
    let data: T?
    let error: String?
}

struct InvokeEnvelope<T: Decodable>: Decodable {
    let data: T?
    let checkoutUrl: String?
    let pendingOAuthUrl: String?
}

struct SessionResponse: Decodable {
    let token: String
    let project: ProjectRef?
}

struct MeResponse: Decodable {
    let project: ProjectRef?
    let projects: [ProjectRef]?
}

struct ProjectRef: Decodable, Identifiable {
    let id: String
    let name: String?
    let isActive: Bool?
}

File: apps/mobile-android/app/src/main/java/com/socialimperialism/mobile/core/ApiClient.kt
package com.socialimperialism.mobile.core

import com.socialimperialism.mobile.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

class ApiException(message: String) : Exception(message)

class ApiClient(private val auth: AuthStore) {
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .build()

    private val jsonType = "application/json".toMediaType()

    suspend fun login(email: String, password: String): SessionResponse = withContext(Dispatchers.IO) {
        val body = JSONObject().put("email", email).put("password", password)
        decodeSession(post("/api/auth/login", body, auth = false))
    }

    suspend fun register(email: String, password: String, name: String?): SessionResponse = withContext(Dispatchers.IO) {
        val body = JSONObject().put("email", email).put("password", password)
        if (!name.isNullOrBlank()) body.put("name", name)
        decodeSession(post("/api/auth/register", body, auth = false))
    }

    suspend fun me(): MeResponse = withContext(Dispatchers.IO) {
        val json = JSONObject(get("/api/auth/me", auth = true))
        MeResponse(
            project = json.optJSONObject("project")?.let { decodeProject(it) },
            projects = json.optJSONArray("projects")?.let { arr ->
                (0 until arr.length()).map { decodeProject(arr.getJSONObject(it)) }
            },
        )
    }

    suspend inline fun <reified T> invoke(channel: String, args: List<Any> = emptyList(), noinline onCheckout: ((String) -> Unit)? = null): T = withContext(Dispatchers.IO) {
        val argsJson = JSONArray()
        args.forEach { arg ->
            when (arg) {
                is Map<*, *> -> argsJson.put(mapToJson(arg))
                else -> argsJson.put(arg)
            }
        }
        val body = JSONObject().put("args", argsJson)
        val json = JSONObject(post("/api/invoke/$channel", body, auth = true))
        val data = json.opt("data")
        if (data is JSONObject) {
            data.optString("checkoutUrl").takeIf { it.isNotBlank() }?.let { onCheckout?.invoke(it) }
        }
        when (T::class) {
            String::class -> (when (data) {
                is String -> data
                else -> data?.toString() ?: ""
            }) as T
            DashboardStats::class -> decodeDashboard(data) as T
            VaultResponse::class -> decodeVault(data) as T
            LoadPromptResponse::class -> decodeLoadPrompt(data) as T
            String::class -> (data?.toString() ?: "") as T
            BillingPlan::class -> decodeBilling(data) as T
            CheckoutResult::class -> decodeCheckout(data) as T
            PublishResult::class -> decodePublish(data) as T
            else -> {
                when (data) {
                    is JSONArray -> {
                        when (T::class) {
                            List::class -> decodeEngagementList(data) as T
                            else -> throw ApiException("Unsupported list type")
                        }
                    }
                    else -> throw ApiException("Unsupported invoke decode: ${T::class}")
                }
            }
        }
    }

    suspend fun invokeLinkedAccounts(): List<LinkedAccount> = withContext(Dispatchers.IO) {
        val json = JSONObject(post("/api/invoke/get-linked-accounts", JSONObject().put("args", JSONArray()), auth = true))
        val data = json.optJSONArray("data") ?: JSONArray()
        (0 until data.length()).map { i ->
            val o = data.getJSONObject(i)
            LinkedAccount(o.getString("id"), o.optString("platform"), o.optString("name"), o.optString("handle"))
        }
    }

    suspend fun invokeEngagementQueue(): List<EngagementQueueItem> = withContext(Dispatchers.IO) {
        val json = JSONObject(post("/api/invoke/get-engagement-queue", JSONObject().put("args", JSONArray()), auth = true))
        val data = json.optJSONArray("data") ?: JSONArray()
        decodeEngagementList(data)
    }

    suspend fun invokeRaw(channel: String, args: List<Any> = emptyList()): JSONObject = withContext(Dispatchers.IO) {
        val argsJson = JSONArray()
        args.forEach { if (it is Map<*, *>) argsJson.put(mapToJson(it)) else argsJson.put(it) }
        JSONObject(post("/api/invoke/$channel", JSONObject().put("args", argsJson), auth = true))
    }

    private fun get(path: String, auth: Boolean): String {
        val req = baseRequest(path, auth).get().build()
        return execute(req, auth, path, null)
    }

    private fun post(path: String, body: JSONObject, auth: Boolean): String {
        val req = baseRequest(path, auth)
            .post(body.toString().toRequestBody(jsonType))
            .build()
        return execute(req, auth, path, body)
    }

    private fun baseRequest(path: String, auth: Boolean): Request.Builder {
        val builder = Request.Builder()
            .url(BuildConfig.API_BASE + path)
            .header("Content-Type", "application/json")
        if (auth) {
            val token = auth.token.value ?: throw ApiException("Not authenticated")
            builder.header("Authorization", "Bearer $token")
            auth.projectId.value?.let { builder.header("x-project-id", it) }
        }
        return builder
    }

    private fun execute(req: Request, auth: Boolean, path: String, retryBody: JSONObject?): String {
        val res = client.newCall(req).execute()
        val text = res.body?.string() ?: "{}"
        if (res.code == 401) {
            auth.clearSession()
            throw ApiException("Session expired")
        }
        if (!res.isSuccessful) {
            val err = try { JSONObject(text).optString("error", res.message) } catch (_: Exception) { res.message }
            if (auth && err.contains("project not found", ignoreCase = true)) {
                auth.setProjectId(null)
                return post(path, retryBody ?: JSONObject(), auth = true)
            }
            throw ApiException(err)
        }
        return text
    }

    private fun decodeSession(text: String): SessionResponse {
        val json = JSONObject(text)
        val token = json.optString("token").ifBlank { json.optJSONObject("data")?.optString("token") ?: "" }
        val project = json.optJSONObject("project") ?: json.optJSONObject("data")?.optJSONObject("project")
        return SessionResponse(token, project?.let { decodeProject(it) })
    }

    private fun decodeProject(o: JSONObject) = ProjectRef(o.getString("id"), o.optString("name"), o.optBoolean("isActive"))

    private fun decodeDashboard(data: Any?): DashboardStats {
        val o = data as? JSONObject ?: JSONObject()
        return DashboardStats(
            totalPosts = o.optInt("totalPosts"),
            aiDrafts = o.optInt("aiDrafts"),
            totalEngagement = o.optInt("totalEngagement"),
            activeKeywords = o.optInt("activeKeywords"),
            linkedAccounts = o.optInt("linkedAccounts"),
            scheduled = o.optInt("scheduled"),
            workerStatus = o.optString("workerStatus"),
            autoRulesEnabled = o.optBoolean("autoRulesEnabled"),
        )
    }

    private fun decodeVault(data: Any?): VaultResponse {
        val o = data as? JSONObject ?: JSONObject()
        val arr = o.optJSONArray("prompts") ?: JSONArray()
        val prompts = (0 until arr.length()).map { i ->
            val p = arr.getJSONObject(i)
            VaultPrompt(p.getString("id"), p.optString("title"), p.optString("body"), p.optString("feature"))
        }
        return VaultResponse(prompts, o.optInt("total"))
    }

    private fun decodeLoadPrompt(data: Any?) = LoadPromptResponse(
        success = (data as? JSONObject)?.optBoolean("success"),
        prompt = (data as? JSONObject)?.optString("prompt"),
        item = null,
    )

    private fun decodeBilling(data: Any?): BillingPlan {
        val o = data as? JSONObject ?: JSONObject()
        return BillingPlan(o.optString("plan"), o.optString("planName"), o.optString("status"), o.optString("priceLabel"), o.optString("billingEmail"))
    }

    private fun decodeCheckout(data: Any?) = CheckoutResult(
        success = (data as? JSONObject)?.optBoolean("success"),
        checkoutUrl = (data as? JSONObject)?.optString("checkoutUrl"),
        error = (data as? JSONObject)?.optString("error"),
    )

    private fun decodePublish(data: Any?) = PublishResult(
        success = (data as? JSONObject)?.optBoolean("success"),
        error = (data as? JSONObject)?.optString("error"),
    )

    private fun decodeEngagementList(arr: JSONArray): List<EngagementQueueItem> =
        (0 until arr.length()).map { i ->
            val o = arr.getJSONObject(i)
            EngagementQueueItem(o.optString("id", "q$i"), o.optString("platform"), o.optString("content"), o.optString("action"), o.optString("status"))
        }

    private fun mapToJson(map: Map<*, *>): JSONObject {
        val o = JSONObject()
        map.forEach { (k, v) -> o.put(k.toString(), v) }
        return o
    }
}